{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting iedb\n",
      "  Downloading iedb-0.0.2-py3-none-any.whl (6.6 kB)\n",
      "Requirement already satisfied: pandas>=1.0.0 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from iedb) (1.2.4)\n",
      "Requirement already satisfied: requests>=2.22.0 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from iedb) (2.25.1)\n",
      "Requirement already satisfied: numpy>=1.16.5 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from pandas>=1.0.0->iedb) (1.21.5)\n",
      "Requirement already satisfied: pytz>=2017.3 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from pandas>=1.0.0->iedb) (2022.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from pandas>=1.0.0->iedb) (2.8.2)\n",
      "Requirement already satisfied: idna<3,>=2.5 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from requests>=2.22.0->iedb) (2.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from requests>=2.22.0->iedb) (2022.6.15)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from requests>=2.22.0->iedb) (1.26.10)\n",
      "Requirement already satisfied: chardet<5,>=3.0.2 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from requests>=2.22.0->iedb) (4.0.0)\n",
      "Requirement already satisfied: six>=1.5 in ./miniconda3/envs/d2l2/lib/python3.8/site-packages (from python-dateutil>=2.7.3->pandas>=1.0.0->iedb) (1.16.0)\n",
      "Installing collected packages: iedb\n",
      "Successfully installed iedb-0.0.2\n"
     ]
    }
   ],
   "source": [
    "!pip install iedb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HLA-A*02:01\n",
      "8\n",
      "HLA-A*02:06\n",
      "8\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/home/maolp/mao/Codeman/Codeproject_log/Sarscov2_iedb.ipynb Cell 2\u001b[0m in \u001b[0;36m<cell line: 15>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2B252/home/maolp/mao/Codeman/Codeproject_log/Sarscov2_iedb.ipynb#ch0000001vscode-remote?line=28'>29</a>\u001b[0m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39m8\u001b[39m,\u001b[39m9\u001b[39m):\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2B252/home/maolp/mao/Codeman/Codeproject_log/Sarscov2_iedb.ipynb#ch0000001vscode-remote?line=29'>30</a>\u001b[0m     \u001b[39mprint\u001b[39m(i)\n\u001b[0;32m---> <a href='vscode-notebook-cell://ssh-remote%2B252/home/maolp/mao/Codeman/Codeproject_log/Sarscov2_iedb.ipynb#ch0000001vscode-remote?line=30'>31</a>\u001b[0m     tcell_res \u001b[39m=\u001b[39m mhci_res \u001b[39m=\u001b[39m  iedb\u001b[39m.\u001b[39;49mquery_mhci_binding(method\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mrecommended\u001b[39;49m\u001b[39m\"\u001b[39;49m, sequence\u001b[39m=\u001b[39;49m\u001b[39mstr\u001b[39;49m(line\u001b[39m.\u001b[39;49mseq), allele\u001b[39m=\u001b[39;49mal, length\u001b[39m=\u001b[39;49mi)\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2B252/home/maolp/mao/Codeman/Codeproject_log/Sarscov2_iedb.ipynb#ch0000001vscode-remote?line=31'>32</a>\u001b[0m     tcell_res[\u001b[39m\"\u001b[39m\u001b[39mname\u001b[39m\u001b[39m\"\u001b[39m]\u001b[39m=\u001b[39mname\n\u001b[1;32m     <a href='vscode-notebook-cell://ssh-remote%2B252/home/maolp/mao/Codeman/Codeproject_log/Sarscov2_iedb.ipynb#ch0000001vscode-remote?line=32'>33</a>\u001b[0m     frames\u001b[39m.\u001b[39mappend(tcell_res)\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/iedb/api.py:50\u001b[0m, in \u001b[0;36mquery_mhci_binding\u001b[0;34m(method, sequence, allele, length)\u001b[0m\n\u001b[1;32m     42\u001b[0m url \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39mhttp://tools-cluster-interface.iedb.org/tools_api/mhci/\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     43\u001b[0m data \u001b[39m=\u001b[39m {\n\u001b[1;32m     44\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mmethod\u001b[39m\u001b[39m'\u001b[39m: method,\n\u001b[1;32m     45\u001b[0m     \u001b[39m'\u001b[39m\u001b[39msequence_text\u001b[39m\u001b[39m'\u001b[39m: sequence, \n\u001b[1;32m     46\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mallele\u001b[39m\u001b[39m'\u001b[39m: allele,\n\u001b[1;32m     47\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mlength\u001b[39m\u001b[39m'\u001b[39m: length\n\u001b[1;32m     48\u001b[0m }\n\u001b[0;32m---> 50\u001b[0m res \u001b[39m=\u001b[39m requests\u001b[39m.\u001b[39;49mpost(\n\u001b[1;32m     51\u001b[0m     url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m     52\u001b[0m     data\u001b[39m=\u001b[39;49mdata\n\u001b[1;32m     53\u001b[0m )\n\u001b[1;32m     55\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m res\u001b[39m.\u001b[39mtext\u001b[39m.\u001b[39mstartswith(\u001b[39m'\u001b[39m\u001b[39mallele\u001b[39m\u001b[39m'\u001b[39m):\n\u001b[1;32m     56\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mAPI returned error: \u001b[39m\u001b[39m{\u001b[39;00mres\u001b[39m.\u001b[39mtext\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/requests/api.py:119\u001b[0m, in \u001b[0;36mpost\u001b[0;34m(url, data, json, **kwargs)\u001b[0m\n\u001b[1;32m    107\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpost\u001b[39m(url, data\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, json\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[1;32m    108\u001b[0m     \u001b[39mr\u001b[39m\u001b[39m\"\"\"Sends a POST request.\u001b[39;00m\n\u001b[1;32m    109\u001b[0m \n\u001b[1;32m    110\u001b[0m \u001b[39m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    116\u001b[0m \u001b[39m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m    117\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 119\u001b[0m     \u001b[39mreturn\u001b[39;00m request(\u001b[39m'\u001b[39;49m\u001b[39mpost\u001b[39;49m\u001b[39m'\u001b[39;49m, url, data\u001b[39m=\u001b[39;49mdata, json\u001b[39m=\u001b[39;49mjson, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/requests/api.py:61\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[39m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[39m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[39m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     60\u001b[0m \u001b[39mwith\u001b[39;00m sessions\u001b[39m.\u001b[39mSession() \u001b[39mas\u001b[39;00m session:\n\u001b[0;32m---> 61\u001b[0m     \u001b[39mreturn\u001b[39;00m session\u001b[39m.\u001b[39;49mrequest(method\u001b[39m=\u001b[39;49mmethod, url\u001b[39m=\u001b[39;49murl, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/requests/sessions.py:542\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    537\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    538\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m'\u001b[39m: timeout,\n\u001b[1;32m    539\u001b[0m     \u001b[39m'\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m'\u001b[39m: allow_redirects,\n\u001b[1;32m    540\u001b[0m }\n\u001b[1;32m    541\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 542\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    544\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/requests/sessions.py:697\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    694\u001b[0m         \u001b[39mpass\u001b[39;00m\n\u001b[1;32m    696\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m stream:\n\u001b[0;32m--> 697\u001b[0m     r\u001b[39m.\u001b[39;49mcontent\n\u001b[1;32m    699\u001b[0m \u001b[39mreturn\u001b[39;00m r\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/requests/models.py:831\u001b[0m, in \u001b[0;36mResponse.content\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    829\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_content \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m    830\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 831\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_content \u001b[39m=\u001b[39m \u001b[39mb\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m.\u001b[39;49mjoin(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49miter_content(CONTENT_CHUNK_SIZE)) \u001b[39mor\u001b[39;00m \u001b[39mb\u001b[39m\u001b[39m'\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    833\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_content_consumed \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    834\u001b[0m \u001b[39m# don't need to release the connection; that's been handled by urllib3\u001b[39;00m\n\u001b[1;32m    835\u001b[0m \u001b[39m# since we exhausted the data.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/requests/models.py:753\u001b[0m, in \u001b[0;36mResponse.iter_content.<locals>.generate\u001b[0;34m()\u001b[0m\n\u001b[1;32m    751\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mhasattr\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw, \u001b[39m'\u001b[39m\u001b[39mstream\u001b[39m\u001b[39m'\u001b[39m):\n\u001b[1;32m    752\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 753\u001b[0m         \u001b[39mfor\u001b[39;00m chunk \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mraw\u001b[39m.\u001b[39mstream(chunk_size, decode_content\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m):\n\u001b[1;32m    754\u001b[0m             \u001b[39myield\u001b[39;00m chunk\n\u001b[1;32m    755\u001b[0m     \u001b[39mexcept\u001b[39;00m ProtocolError \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/urllib3/response.py:575\u001b[0m, in \u001b[0;36mHTTPResponse.stream\u001b[0;34m(self, amt, decode_content)\u001b[0m\n\u001b[1;32m    559\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    560\u001b[0m \u001b[39mA generator wrapper for the read() method. A call will block until\u001b[39;00m\n\u001b[1;32m    561\u001b[0m \u001b[39m``amt`` bytes have been read from the connection or until the\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[39m    'content-encoding' header.\u001b[39;00m\n\u001b[1;32m    573\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    574\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunked \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msupports_chunked_reads():\n\u001b[0;32m--> 575\u001b[0m     \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mread_chunked(amt, decode_content\u001b[39m=\u001b[39mdecode_content):\n\u001b[1;32m    576\u001b[0m         \u001b[39myield\u001b[39;00m line\n\u001b[1;32m    577\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/urllib3/response.py:770\u001b[0m, in \u001b[0;36mHTTPResponse.read_chunked\u001b[0;34m(self, amt, decode_content)\u001b[0m\n\u001b[1;32m    768\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunk_left \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    769\u001b[0m     \u001b[39mbreak\u001b[39;00m\n\u001b[0;32m--> 770\u001b[0m chunk \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_handle_chunk(amt)\n\u001b[1;32m    771\u001b[0m decoded \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_decode(\n\u001b[1;32m    772\u001b[0m     chunk, decode_content\u001b[39m=\u001b[39mdecode_content, flush_decoder\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m\n\u001b[1;32m    773\u001b[0m )\n\u001b[1;32m    774\u001b[0m \u001b[39mif\u001b[39;00m decoded:\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/site-packages/urllib3/response.py:714\u001b[0m, in \u001b[0;36mHTTPResponse._handle_chunk\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    712\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunk_left \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m    713\u001b[0m \u001b[39melif\u001b[39;00m amt \u001b[39m<\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunk_left:\n\u001b[0;32m--> 714\u001b[0m     value \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_fp\u001b[39m.\u001b[39;49m_safe_read(amt)\n\u001b[1;32m    715\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunk_left \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mchunk_left \u001b[39m-\u001b[39m amt\n\u001b[1;32m    716\u001b[0m     returned_chunk \u001b[39m=\u001b[39m value\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/http/client.py:613\u001b[0m, in \u001b[0;36mHTTPResponse._safe_read\u001b[0;34m(self, amt)\u001b[0m\n\u001b[1;32m    606\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_safe_read\u001b[39m(\u001b[39mself\u001b[39m, amt):\n\u001b[1;32m    607\u001b[0m     \u001b[39m\"\"\"Read the number of bytes requested.\u001b[39;00m\n\u001b[1;32m    608\u001b[0m \n\u001b[1;32m    609\u001b[0m \u001b[39m    This function should be used when <amt> bytes \"should\" be present for\u001b[39;00m\n\u001b[1;32m    610\u001b[0m \u001b[39m    reading. If the bytes are truly not available (due to EOF), then the\u001b[39;00m\n\u001b[1;32m    611\u001b[0m \u001b[39m    IncompleteRead exception can be used to detect the problem.\u001b[39;00m\n\u001b[1;32m    612\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 613\u001b[0m     data \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfp\u001b[39m.\u001b[39;49mread(amt)\n\u001b[1;32m    614\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(data) \u001b[39m<\u001b[39m amt:\n\u001b[1;32m    615\u001b[0m         \u001b[39mraise\u001b[39;00m IncompleteRead(data, amt\u001b[39m-\u001b[39m\u001b[39mlen\u001b[39m(data))\n",
      "File \u001b[0;32m~/miniconda3/envs/d2l2/lib/python3.8/socket.py:669\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    667\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m    668\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 669\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sock\u001b[39m.\u001b[39;49mrecv_into(b)\n\u001b[1;32m    670\u001b[0m     \u001b[39mexcept\u001b[39;00m timeout:\n\u001b[1;32m    671\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_timeout_occurred \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import pandas as  pd\n",
    "import datetime\n",
    "# !pip install iedb\n",
    "from Bio import SeqIO\n",
    "import re,sys\n",
    "import iedb\n",
    "# num=8\n",
    "savename=\"Father\"\n",
    "starttime = datetime.datetime.now()\n",
    "frames=[]\n",
    "# allelename=[\"HLA-A*01:01\",\"HLA-A*02:01\",\"HLA-A*02:06\",\"HLA-A*03:01\",\"HLA-A*11:01\",\"HLA-A*23:01\",\"HLA-A*24:02\",\"HLA-A*25:01\",\"HLA-A*26:01\",\"HLA-A*29:02\",\"HLA-A*30:01\"\n",
    "# ,\"HLA-A*31:01\",\"HLA-A*32:01\",\"HLA-A*33:03\",\"HLA-A*68:01\",\"HLA-A*68:02\",\"HLA-A*74:01\"]\n",
    "allelename=[\"HLA-A*02:01\",\"HLA-A*02:06\",\"HLA-A*23:01\",\"HLA-A*24:02\",\"HLA-A*30:01\"]\n",
    "seq=SeqIO.parse(\"/home/maolp/data3/Project/Sarscov2_iedb/refsars.fa\", \"fasta\")\n",
    "for line in seq:\n",
    "    name=line.id.split(\"&&\")[1]\n",
    "    # # print(name)\n",
    "    # if re.findall(\"surface\",line.id):\n",
    "    #     # print(line.id)\n",
    "    #     print(name)\n",
    "\n",
    "    #     print(line.seq.rstrip())\n",
    "        \n",
    "\n",
    "    #     # print(iedb.predict_antigen_sequence(line.seq))\n",
    "    #     print(\"-----------------------------------------------------\")\n",
    "    for al in allelename:\n",
    "        print(al)\n",
    "        for i in range(8,9):\n",
    "            print(i)\n",
    "            tcell_res = mhci_res =  iedb.query_mhci_binding(method=\"recommended\", sequence=str(line.seq), allele=al, length=i)\n",
    "            tcell_res[\"name\"]=name\n",
    "            frames.append(tcell_res)\n",
    "            \n",
    "\n",
    "result_sp = pd.concat(frames)\n",
    "\n",
    "result_sp.to_csv(\"/home/maolp/data3/Project/Sarscov2_iedb/refsar_8_pep_result.csv\",index=True)\n",
    "endtime = datetime.datetime.now()\n",
    "print(\"time:\",endtime - starttime)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as  pd\n",
    "import datetime\n",
    "# !pip install iedb\n",
    "from Bio import SeqIO\n",
    "import re,sys\n",
    "import iedb\n",
    "\n",
    "def seq_iedb(num,savefile):\n",
    "    num=8\n",
    "    savename=\"Father\"\n",
    "    starttime = datetime.datetime.now()\n",
    "    frames=[]\n",
    "    # allelename=[\"HLA-A*01:01\",\"HLA-A*02:01\",\"HLA-A*02:06\",\"HLA-A*03:01\",\"HLA-A*11:01\",\"HLA-A*23:01\",\"HLA-A*24:02\",\"HLA-A*25:01\",\"HLA-A*26:01\",\"HLA-A*29:02\",\"HLA-A*30:01\"\n",
    "    # ,\"HLA-A*31:01\",\"HLA-A*32:01\",\"HLA-A*33:03\",\"HLA-A*68:01\",\"HLA-A*68:02\",\"HLA-A*74:01\"]\n",
    "    allelename=[\"HLA-A*02:01\",\"HLA-A*02:06\",\"HLA-A*23:01\",\"HLA-A*24:02\",\"HLA-A*30:01\"]\n",
    "    seq=SeqIO.parse(\"/home/maolp/data3/Project/Sarscov2_iedb/refsars.fa\", \"fasta\")\n",
    "    for line in seq:\n",
    "        name=line.id.split(\"&&\")[1]\n",
    "        # # print(name)\n",
    "        # if re.findall(\"surface\",line.id):\n",
    "        #     # print(line.id)\n",
    "        #     print(name)\n",
    "\n",
    "        #     print(line.seq.rstrip())\n",
    "            \n",
    "\n",
    "        #     # print(iedb.predict_antigen_sequence(line.seq))\n",
    "        #     print(\"-----------------------------------------------------\")\n",
    "        for al in allelename:\n",
    "            print(al)\n",
    "            for i in range(num,num+1):\n",
    "                print(i)\n",
    "                tcell_res = mhci_res =  iedb.query_mhci_binding(method=\"recommended\", sequence=str(line.seq), allele=al, length=i)\n",
    "                tcell_res[\"name\"]=name\n",
    "                frames.append(tcell_res)\n",
    "                \n",
    "\n",
    "    result_sp = pd.concat(frames)\n",
    "\n",
    "    result_sp.to_csv(\"/home/maolp/data3/Project/Sarscov2_iedb/refsar_\"+num+\"_pep_result.csv\",index=True)\n",
    "    endtime = datetime.datetime.now()\n",
    "    print(\"time:\",endtime - starttime)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import iedb\n",
    "\n",
    "# Send POST request to MHC class I peptide binding prediction tool:\n",
    "# mhci_res = iedb.query_mhci_binding(method=\"recommended\", sequence=\"ARFTGIKTA\", allele=\"HLA-A*02:01\", length=None)\n",
    "\n",
    "# # Send POST request to MHC class II peptide binding prediction tool:\n",
    "# mhcii_res = iedb.query_mhcii_binding(method=\"nn_align\", sequence=\"SLYNTVATLYCVHQRIDV\", allele=\"HLA-DRB1*01:01\", length=None)\n",
    "\n",
    "# # Send POST request to T-cell epitope prediction tool:\n",
    "tcell_res = iedb.query_tcell_epitope(method=\"recommended\", sequence=\"SLYNTVATLYCVHQRIDV\", allele=\"HLA-A*01:01\", length=9, proteasome='immuno')\n",
    "\n",
    "# # Send POST request to peptide prediction tool:\n",
    "# pep_res = iedb.query_peptide_prediction(method=\"mhcnp\", sequence=\"SLYNTVATLYCVHQRIDV\", allele=\"HLA-A*02:01\", length=9)\n",
    "\n",
    "# # Send POST request to B-cell epitope prediction tool:\n",
    "# bcell_res = iedb.query_bcell_epitope(method=\"Emini\", sequence=\"VLSEGEWQLVLHVWAKVEADVAGHGQDILIRLFKSHPETLEKFDRFKHLKTE\", window_size=9)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Query prediction tool used to produce overall score for each peptide's \n",
      "    intrinsic potential of being a T cell epitope.\n",
      "    \n",
      "    Algorithm combines predictors of proteasomal processing, TAP transport, \n",
      "    and MHC binding to produce score.\n",
      "\n",
      "    IEDB official web tool can be found here: http://tools.iedb.org/processing/\n",
      "\n",
      "            Parameters:\n",
      "                    method (str): Prediction method. Available argument \n",
      "                        options:\n",
      "                            - recommended\n",
      "                            - netmhcpan\n",
      "                            - ann\n",
      "                            - smmpmbec\n",
      "                            - smm\n",
      "                            - comblib_sidney2008\n",
      "                            - netmhccons\n",
      "                            - pickpocket\n",
      "                        More information on prediction methods can be found \n",
      "                        here: http://tools.iedb.org/processing/help/#Method\n",
      "\n",
      "                    sequence ([str, list]): Peptide amino acid sequence or\n",
      "                        list of sequences.\n",
      "\n",
      "                    allele ([str, list]): Single or multiple HLA alleles. \n",
      "                        Provide single alleles as a string. Multiple alleles\n",
      "                        should be formatted in a comma separated list. Example\n",
      "                        below:\n",
      "                            Ex.) HLA-A*02:01\n",
      "\n",
      "                    length ([int, list]): Peptide length or list of lengths.\n",
      "                        Possible lengths:\n",
      "                            - 8, 9, 10, 11, 12, 13, 14\t\n",
      "\n",
      "                    proteasome (str): Must be either 'immuno' or 'constitutive'\n",
      "                            Options reference the two types of proteasomes, the \n",
      "                            constitutively expressed 'house-keeping' type, \n",
      "                            and immuno proteasomes that are induced by IFN-Î³ \n",
      "                            secretion.\n",
      "\n",
      "            Returns:\n",
      "                    dataframe (pandas.DataFrame): Tabular results formatted\n",
      "                        as pandas.DataFrame\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "print(iedb.query_tcell_epitope.__doc__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele</th>\n",
       "      <th>seq_num</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>peptide</th>\n",
       "      <th>proteasome_score</th>\n",
       "      <th>tap_score</th>\n",
       "      <th>mhc_score</th>\n",
       "      <th>processing_score</th>\n",
       "      <th>total_score</th>\n",
       "      <th>ic50_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>LYNTVATLY</td>\n",
       "      <td>1.2896</td>\n",
       "      <td>1.3692</td>\n",
       "      <td>-3.8196</td>\n",
       "      <td>2.6588</td>\n",
       "      <td>-1.1608</td>\n",
       "      <td>6600.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>SLYNTVATL</td>\n",
       "      <td>1.5479</td>\n",
       "      <td>0.5087</td>\n",
       "      <td>-4.4381</td>\n",
       "      <td>2.0566</td>\n",
       "      <td>-2.3814</td>\n",
       "      <td>27418.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>ATLYCVHQR</td>\n",
       "      <td>1.1083</td>\n",
       "      <td>0.7256</td>\n",
       "      <td>-4.4730</td>\n",
       "      <td>1.8338</td>\n",
       "      <td>-2.6392</td>\n",
       "      <td>29717.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>NTVATLYCV</td>\n",
       "      <td>1.0686</td>\n",
       "      <td>0.1535</td>\n",
       "      <td>-4.0919</td>\n",
       "      <td>1.2221</td>\n",
       "      <td>-2.8698</td>\n",
       "      <td>12356.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "      <td>TLYCVHQRI</td>\n",
       "      <td>1.0977</td>\n",
       "      <td>0.3789</td>\n",
       "      <td>-4.4056</td>\n",
       "      <td>1.4766</td>\n",
       "      <td>-2.9290</td>\n",
       "      <td>25444.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>TVATLYCVH</td>\n",
       "      <td>1.0610</td>\n",
       "      <td>-0.1120</td>\n",
       "      <td>-4.1893</td>\n",
       "      <td>0.9490</td>\n",
       "      <td>-3.2402</td>\n",
       "      <td>15462.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>VATLYCVHQ</td>\n",
       "      <td>1.3263</td>\n",
       "      <td>-0.0089</td>\n",
       "      <td>-4.5738</td>\n",
       "      <td>1.3174</td>\n",
       "      <td>-3.2564</td>\n",
       "      <td>37477.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>YCVHQRIDV</td>\n",
       "      <td>0.9107</td>\n",
       "      <td>0.1898</td>\n",
       "      <td>-4.5393</td>\n",
       "      <td>1.1005</td>\n",
       "      <td>-3.4388</td>\n",
       "      <td>34621.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>YNTVATLYC</td>\n",
       "      <td>0.9997</td>\n",
       "      <td>-0.0638</td>\n",
       "      <td>-4.3793</td>\n",
       "      <td>0.9359</td>\n",
       "      <td>-3.4434</td>\n",
       "      <td>23948.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-A*01:01</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "      <td>LYCVHQRID</td>\n",
       "      <td>1.0055</td>\n",
       "      <td>-0.6641</td>\n",
       "      <td>-4.6692</td>\n",
       "      <td>0.3414</td>\n",
       "      <td>-4.3279</td>\n",
       "      <td>46691.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        allele seq_num start end length    peptide proteasome_score tap_score  \\\n",
       "0  HLA-A*01:01       1     2  10      9  LYNTVATLY           1.2896    1.3692   \n",
       "1  HLA-A*01:01       1     1   9      9  SLYNTVATL           1.5479    0.5087   \n",
       "2  HLA-A*01:01       1     7  15      9  ATLYCVHQR           1.1083    0.7256   \n",
       "3  HLA-A*01:01       1     4  12      9  NTVATLYCV           1.0686    0.1535   \n",
       "4  HLA-A*01:01       1     8  16      9  TLYCVHQRI           1.0977    0.3789   \n",
       "5  HLA-A*01:01       1     5  13      9  TVATLYCVH           1.0610   -0.1120   \n",
       "6  HLA-A*01:01       1     6  14      9  VATLYCVHQ           1.3263   -0.0089   \n",
       "7  HLA-A*01:01       1    10  18      9  YCVHQRIDV           0.9107    0.1898   \n",
       "8  HLA-A*01:01       1     3  11      9  YNTVATLYC           0.9997   -0.0638   \n",
       "9  HLA-A*01:01       1     9  17      9  LYCVHQRID           1.0055   -0.6641   \n",
       "\n",
       "  mhc_score processing_score total_score ic50_score  \n",
       "0   -3.8196           2.6588     -1.1608     6600.4  \n",
       "1   -4.4381           2.0566     -2.3814    27418.9  \n",
       "2   -4.4730           1.8338     -2.6392    29717.4  \n",
       "3   -4.0919           1.2221     -2.8698    12356.4  \n",
       "4   -4.4056           1.4766     -2.9290    25444.0  \n",
       "5   -4.1893           0.9490     -3.2402    15462.1  \n",
       "6   -4.5738           1.3174     -3.2564    37477.9  \n",
       "7   -4.5393           1.1005     -3.4388    34621.1  \n",
       "8   -4.3793           0.9359     -3.4434    23948.1  \n",
       "9   -4.6692           0.3414     -4.3279    46691.7  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcell_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('d2l2': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "8f2a546bff4b941b65124b22fc880a785f84f5d44aa036743b446d26a6d98be5"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
